### ISSUE TITLE
Stake pools have weirdly high performance

### ISSUE THREAD

**Issue by Anviking** [cardano-foundation/cardano-wallet#1327]
# Context
 
We expect statistics to cause small pools to have high performance with small amount of data.

| Information  | -                                                     |
| ---          | ---                                                   |
| Version      | 2020.1.27  |
| Platform     | Mac OS |
| Installation | Internal build / from source |

# Steps to Reproduce

```bash
cardano-wallet-jormungandr stake-pool list | jq "." > response.json
```

Here is my sqlite db:
[stake-pools.sqlite.zip](https://github.com/input-output-hk/cardano-wallet/files/4160937/stake-pools.sqlite.zip)

## Expected behavior
Unclear. Supposedly there should be few pools with a performance of over 100%.

## Actual behavior

There is a top pool with 156 produced blocks, and performance of 325%.

(exact json from another build, but should be similar)
```json
{
    "saturation": 0.010251718287484165,
    "metrics": {
      "controlled_stake": {
        "quantity": 1160630473392,
        "unit": "lovelace"
      },
      "produced_blocks": {
        "quantity": 156,
        "unit": "block"
      }
    },
    "cost": {
      "quantity": 250000000,
      "unit": "lovelace"
    },
    "margin": {
      "quantity": 5,
      "unit": "percent"
    },
    "apparent_performance": 3.140912576100334,
    "metadata": {
      "homepage": "https://www.archpool.io",
      "owner": "ed25519_pk1th5pgvs3cg7qjwq5gw8rmf4jh6fdpckp37flpjuxfzv75gkfvr3syxxcpm",
      "name": "ARCH Staking Pool",
      "ticker": "ARCH1",
      "pledge_address": "addr1s4w7s9pjz8prczfcz3pcu0dxk2lf958zcx8e8uxtseyfn63ze9swxug37z7",
      "description": "Archimedes Staking Pool for ADA Cardano"
    },
    "id": "08b46d4bc5d8110c76dde2cc7a4c9dda0830c4434dedd7fcd32b1a4815c1a4f3",
    "desirability": 102767222060.36679
  }
```

---

# Resolution 

1. Improve logging in `Performance` module.
2. Follow the spec more strictly on the performance calculation

---

# QA

**Comment by Anviking** [comment]
Here's the pool: https://pooltool.io/pool/08b46d4bc5d8110c76dde2cc7a4c9dda0830c4434dedd7fcd32b1a4815c1a4f3

**Comment by Anviking** [comment]
Preliminary logs showing the epochs we use, and performances calculated using each of them. (while also following the spec more closely #1329).

It would be nice to have the epoch numbers also. Maybe we have been unlucky and stored stake distributions for the very epochs where the pool did unusually well.
```
Epoch stats[
 poolProduction = 2
 poolStake = 1126660897267
 totalStake = 10531492964919807
 resulting relativeStake = 1.0698016900546626e-4
 totalProduction = 2915
], perf: 6.413397481627157
[cardano-wallet.pools-engine:Info:38] [2020-02-05 16:41:54.63 UTC] Epoch stats[
 poolProduction = 2
 poolStake = 1127468807540
 totalStake = 10550100343406315
 resulting relativeStake = 1.0686806483737894e-4
 totalProduction = 3031
], perf: 6.1744192328181695
[cardano-wallet.pools-engine:Info:38] [2020-02-05 16:41:54.67 UTC] Epoch stats[
 poolProduction = 1
 poolStake = 1127468807540
 totalStake = 10586905884510435
 resulting relativeStake = 1.0649653636664372e-4
 totalProduction = 3102
], perf: 3.0270718071838303
[cardano-wallet.pools-engine:Info:24] [2020-02-05 16:41:54.74 UTC] Applying blocks [53.38648 ... 53.38648]
[cardano-wallet.pools-engine:Info:24] [2020-02-05 16:41:54.74 UTC] Writing stake-distribution for epoch 53
[cardano-wallet.pools-engine:Info:38] [2020-02-05 16:41:54.82 UTC] Epoch stats[
 poolProduction = 0
 poolStake = 1159797306717
 totalStake = 10673086444132557
 resulting relativeStake = 1.0866559666576946e-4
 totalProduction = 3020
], perf: 0.0
[cardano-wallet.pools-engine:Info:38] [2020-02-05 16:41:55.01 UTC] Epoch stats[
 poolProduction = 0
 poolStake = 1160630473392
 totalStake = 11321326248390560
 resulting relativeStake = 1.0251718287484165e-4
 totalProduction = 3011
], perf: 0.0
```

**Comment by Anviking** [comment]
New logs (with epoch numbers this time):

```
Stored statistics for the ARCH1 pool:

[cardano-wallet.pools-engine:Info:104] [2020-02-06 08:13:45.05 UTC] epoch:40Epoch stats
 poolProduction = 2
 poolStake = 1127468807540
 totalStake = 10550100343406315
 resulting relativeStake = 1.0686806483737894e-4
 totalProduction = 3031
, perf: 6.1744192328181695
[cardano-wallet.pools-engine:Info:104] [2020-02-06 08:13:45.09 UTC] epoch:41Epoch stats
 poolProduction = 1
 poolStake = 1127468807540
 totalStake = 10586905884510435
 resulting relativeStake = 1.0649653636664372e-4
 totalProduction = 3102
, perf: 3.0270718071838303
[cardano-wallet.pools-engine:Info:104] [2020-02-06 08:13:45.24 UTC] epoch:45Epoch stats
 poolProduction = 0
 poolStake = 1159797306717
 totalStake = 10673086444132557
 resulting relativeStake = 1.0866559666576946e-4
 totalProduction = 3020
, perf: 0.0
[cardano-wallet.pools-engine:Info:104] [2020-02-06 08:13:45.44 UTC] epoch:53Epoch stats
 poolProduction = 0
 poolStake = 1160630473392
 totalStake = 11321326248390560
 resulting relativeStake = 1.0251718287484165e-4
 totalProduction = 3364
, perf: 0.0
[cardano-wallet.pools-engine:Info:104] [2020-02-06 08:13:45.45 UTC] epoch:54Epoch stats
 poolProduction = 0
 poolStake = 1164200542537
 totalStake = 11352044149698991
 resulting relativeStake = 1.0255426486936889e-4
 totalProduction = 527
, perf: 0.0

```

**Comment by Anviking** [comment]
To me it seems like nothing blatantly is wrong.

More epochs of stake-distributions would help, of course.

Switching back to `slots in epoch * f` instead of `actual number of total blocks produced this epoch`, could be an option, as (*I think* 🤔 ) suggested by [Philipp and Duncan](https://input-output-rnd.slack.com/archives/GJY4XC3DM/p1580899029067400?thread_ts=1580887647.056900&cid=GJY4XC3DM).

**Comment by KtorZ** [comment]
As node client, we do not have access to past stake distributions.
Meaning that, we cannot compute the apparent performance as described in the paper and this is precisely why we had to deviate from the spec and adjust things a bit. Because we can't reliably know stake distribution of the past epoch (which is over and  "stable"), we compute the performance based on the ongoing epoch which is.. well, ongoing. Therefore, the approximation is even worst for luck is even more important in the calculation (if you're elected early, you'll look like you're performing better).
There's no silver bullet to this problem. Different ways of computing the performance on the ongoing epoch have different trade-offs. We've tried several, and I am convinced that the latest one is the less worse of all. 300% is not completely insane for a pool that is apparently performing very well compared to the expectation. (edited)
